thetis/rungekutta.py

```diff
@@ -0,0 +1,949 @@
+
+    def solve_stage(self, i_stage, t, update_forcings=None):
+        if self._nontrivial:
+            if update_forcings is not None:
+                update_forcings(t + self.c[i_stage]*self.dt)
+
+            if i_stage == 0:
+                self.stage_sol[0].assign(self.solution)
+
+            # solve tendency
+            self.solver.solve()
+
+            # solve the next internal solution
+            # BUG starting with sol_sum = 0 results in different solution!
+            sol_sum = float(self.beta[i_stage + 1][i_stage])*self.tendency
+            for j in range(i_stage + 1):
+                sol_sum += float(self.alpha[i_stage + 1][j])*self.stage_sol[j]
+            self.solution.assign(sol_sum)
+            if i_stage < self.n_stages - 1:
+                self.stage_sol[i_stage + 1].assign(self.solution)
+
+    def advance(self, t, update_forcings=None):
+        """Advances equations for one time step."""
+        for i in xrange(self.n_stages):
+            self.solve_stage(i, t, update_forcings)
+
+
+class ERKGenericALE2(RungeKuttaTimeIntegrator):
+    """
+    Generic explicit Runge-Kutta time integrator for conservative ALE schemes.
+
+    Implements the Butcher tableau.
+    """
+
+    def __init__(self, equation, solution, fields, dt, bnd_conditions=None, solver_parameters={}):
+        """Creates forms for the time integrator"""
+        super(ERKGenericALE2, self).__init__(equation, solution, fields, dt, solver_parameters)
+
+        self.l_form = Function(self.equation.function_space, name='linear form')
+        self.msol_old = Function(self.equation.function_space, name='old dual solution')
+        self.stage_mk = []  # mass_matrix*tendency
+        for i in range(self.n_stages):
+            s = Function(self.equation.function_space, name='dual tendency {:}'.format(i))
+            self.stage_mk.append(s)
+
+        # fully explicit evaluation
+        self.a_rk = self.equation.mass_term(self.equation.trial)
+        self.l_rk = self.dt_const*self.equation.residual('all', self.solution, self.solution, self.fields, self.fields, bnd_conditions)
+        self.mass_term = self.equation.mass_term(self.solution)
+
+        self._nontrivial = self.l_rk != 0
+
+        self.update_solver()
+
+    def update_solver(self):
+        if self._nontrivial:
+            self.A = assemble(self.a_rk)
+
+    def initialize(self, solution):
+        """Assigns initial conditions to all required fields."""
+        assemble(self.mass_term, self.msol_old)
+
+    def update_solution(self, i_stage):
+        """
+        Updates solution to i_stage sub-stage.
+
+        Tendencies must have been evaluated first.
+        """
+        if self._nontrivial:
+            # construct full form: L = c*dt*F + M_n*sol_n + ...
+            # everything is assembled, just sum up functions
+            self.l_form.assign(self.msol_old)
+            for j in range(i_stage):
+                self.l_form += float(self.a[i_stage][j])*self.stage_mk[j]
```

Comment:
Here I would write:

```python
self.l_form.assign(self.msol_old + reduce(operator.add, map(operator.mul, self.a[i_stage][:i_stage], self.stage_mk[:i_stage])))
```

So you do one loop, rather than 1 + i_stage loops.